### PR TITLE
Add referenced libraries as reference container

### PR DESCRIPTION
**Pull request by Flanker32**
Add referenced libraries as reference container

### CODE REVIEW

File: jdtls.ext/com.microsoft.jdtls.ext.core/src/com/microsoft/jdtls/ext/core/PackageCommand.java [hunk 1]
```diff
@@ -174,36 +181,66 @@
...
-                    return null;
-                }).filter(containerNode -> containerNode != null).collect(Collectors.toList());
+                List<PackageNode> result = Arrays.stream(references).filter(entry -> entry.getEntryKind() != IClasspathEntry.CPE_LIBRARY)
+                        .map(entry -> getNodeFromClasspathEntry(entry, javaProject, NodeKind.CONTAINER)).filter(containerNode -> containerNode != null)
+                        .collect(Collectors.toList());
+
+                List<IClasspathEntry> referenceLibraries = Arrays.stream(references).filter(entry -> entry.getEntryKind() == IClasspathEntry.CPE_LIBRARY)
+                        .collect(Collectors.toList());
+                if (referenceLibraries.size() > 0) {
+                    result.add(new ContainerNode("ReferencedLibraries", REFERENCED_LIBRARIES_PATH, NodeKind.CONTAINER, IClasspathEntry.CPE_CONTAINER));
```
Comment (andxu) [inline jdtls.ext/com.microsoft.jdtls.ext.core/src/com/microsoft/jdtls/ext/core/PackageCommand.java]: ContainerNode("ReferencedLibraries", REFERENCED_LIBRARIES_PATH, NodeKind.CONTAINER, IClasspathEntry.CPE_CONTAINER) should be a constant

File: jdtls.ext/com.microsoft.jdtls.ext.core/src/com/microsoft/jdtls/ext/core/PackageCommand.java [hunk 2]
```diff
@@ -154,8 +156,13 @@
             if (typeRoot instanceof IClassFile) {
                 IClasspathEntry entry = pkgRoot.getRawClasspathEntry();
                 IClasspathContainer container = JavaCore.getClasspathContainer(entry.getPath(), typeRoot.getJavaProject());
-                PackageNode containerNode = new ContainerNode(container.getDescription(), container.getPath().toPortableString(), NodeKind.CONTAINER,
-                        entry.getEntryKind());
+                PackageNode containerNode = null;
+                if (entry.getEntryKind() == IClasspathEntry.CPE_LIBRARY) {
```
Comment (andxu) [inline jdtls.ext/com.microsoft.jdtls.ext.core/src/com/microsoft/jdtls/ext/core/PackageCommand.java]: should consider/test the scenario of IClasspathEntry.CPE_VARIABLE

File: jdtls.ext/com.microsoft.jdtls.ext.core/src/com/microsoft/jdtls/ext/core/PackageCommand.java [hunk 3]
```diff
@@ -174,36 +190,80 @@
...
+            // Ignore it
+        }
+        return null;
+    }
+
+    private static PackageNode getNodeFromClasspathVariable(IClasspathEntry classpathEntry) {
+        IClasspathEntry entry = JavaCore.getResolvedClasspathEntry(classpathEntry);
+        String name = classpathEntry.getPath().toPortableString();
+        String path = entry.getPath().toPortableString();
+        if (new File(path).isDirectory()) {
```
Comment (andxu) [inline jdtls.ext/com.microsoft.jdtls.ext.core/src/com/microsoft/jdtls/ext/core/PackageCommand.java]: merge to remove duplicate code

File: jdtls.ext/com.microsoft.jdtls.ext.core/src/com/microsoft/jdtls/ext/core/PackageCommand.java [hunk 4]
```diff
@@ -174,36 +190,80 @@
...
+                    // Use package name as package root name
+                    String[] pathSegments = container.getPath().segments();
+                    return new PackageRootNode(pathSegments[pathSegments.length - 1], container.getPath().toPortableString(), nodeKind,
+                            IClasspathEntry.CPE_PROJECT);
+                default:
+                    return null;
+                }
+            }
+        } catch (CoreException e) {
+            e.printStackTrace();
```
Comment (andxu) [inline jdtls.ext/com.microsoft.jdtls.ext.core/src/com/microsoft/jdtls/ext/core/PackageCommand.java]: JdtlsExtActivator.logException("xxxx ", e);

File: jdtls.ext/com.microsoft.jdtls.ext.core/src/com/microsoft/jdtls/ext/core/PackageCommand.java [hunk 5]
```diff
@@ -174,36 +190,80 @@
...
+    /**
+     * Get the correspond node of classpath, it may be container or a package root
+     *
+     * @param classpathEntry
+     *            classpath
+     * @param javaProject
+     *            correspond java project
+     * @param nodeKind
+     *            could be CONTAINER or PACKAGEROOT(for referenced libraries)
+     * @return
```
Comment (andxu) [inline jdtls.ext/com.microsoft.jdtls.ext.core/src/com/microsoft/jdtls/ext/core/PackageCommand.java]: return doc?

File: jdtls.ext/com.microsoft.jdtls.ext.core/src/com/microsoft/jdtls/ext/core/PackageCommand.java [hunk 6]
```diff
@@ -232,6 +291,16 @@
                         }
                     }
                     return children;
+                } else if (query.getPath().equals(REFERENCED_LIBRARIES_PATH)) {
+                    // Process referenced libraries
+                    List<PackageNode> referLibs = Arrays.stream(references).filter(entry -> entry.getEntryKind() == IClasspathEntry.CPE_LIBRARY)
+                            .map(classpath -> getNodeFromClasspathEntry(classpath, javaProject, NodeKind.PACKAGEROOT)).filter(entry -> entry != null)
+                            .collect(Collectors.toList());
+                    List<PackageNode> referValues = Arrays.stream(references).filter(entry -> entry.getEntryKind() == IClasspathEntry.CPE_VARIABLE)
```
Comment (andxu) [inline jdtls.ext/com.microsoft.jdtls.ext.core/src/com/microsoft/jdtls/ext/core/PackageCommand.java]: referVariables

File: jdtls.ext/com.microsoft.jdtls.ext.core/src/com/microsoft/jdtls/ext/core/PackageCommand.java [hunk 7]
```diff
@@ -174,36 +190,80 @@
...
+
+    private static PackageNode getNodeFromClasspathVariable(IClasspathEntry classpathEntry) {
+        IClasspathEntry entry = JavaCore.getResolvedClasspathEntry(classpathEntry);
+        String name = classpathEntry.getPath().toPortableString();
+        String path = entry.getPath().toPortableString();
+        if (new File(path).isDirectory()) {
+            return new PackageRootNode(name, path, NodeKind.PACKAGEROOT, IClasspathEntry.CPE_LIBRARY);
+        } else {
+            return new PackageRootNode(name, path, NodeKind.PACKAGEROOT, IClasspathEntry.CPE_PROJECT);
+        }
```
Comment (andxu) [inline jdtls.ext/com.microsoft.jdtls.ext.core/src/com/microsoft/jdtls/ext/core/PackageCommand.java]: why IClasspathEntry.CPE_PROJECT ?

Comment (Flanker32) [reply jdtls.ext/com.microsoft.jdtls.ext.core/src/com/microsoft/jdtls/ext/core/PackageCommand.java]: Follow by the behavior of Maven references, jar was set to IClasspathEntry.CPE_PROJECT

File: jdtls.ext/com.microsoft.jdtls.ext.core/src/com/microsoft/jdtls/ext/core/PackageCommand.java [hunk 8]
```diff
@@ -154,8 +164,14 @@
             if (typeRoot instanceof IClassFile) {
                 IClasspathEntry entry = pkgRoot.getRawClasspathEntry();
                 IClasspathContainer container = JavaCore.getClasspathContainer(entry.getPath(), typeRoot.getJavaProject());
-                PackageNode containerNode = new ContainerNode(container.getDescription(), container.getPath().toPortableString(), NodeKind.CONTAINER,
-                        entry.getEntryKind());
+                PackageNode containerNode = null;
+                if (entry.getEntryKind() == IClasspathEntry.CPE_LIBRARY || entry.getEntryKind() == IClasspathEntry.CPE_VARIABLE) {
+                    containerNode = new ContainerNode(REFERENCED_LIBRARIES_CONTAINER_NAME, REFERENCED_LIBRARIES_PATH, NodeKind.CONTAINER,
```
Comment (andxu) [inline jdtls.ext/com.microsoft.jdtls.ext.core/src/com/microsoft/jdtls/ext/core/PackageCommand.java]: ontainerNode = new ContainerNode(REFERENCED_LIBRARIES_CONTAINER_NAME, REFERENCED_LIBRARIES_PATH, NodeKind.CONTAINER, duplicate with line 204

File: jdtls.ext/com.microsoft.jdtls.ext.core/src/com/microsoft/jdtls/ext/core/PackageCommand.java [hunk 9]
```diff
@@ -174,36 +190,83 @@
...
-                    } catch (CoreException e) {
-                        // Ignore it
-                    }
-                    return null;
-                }).filter(containerNode -> containerNode != null).collect(Collectors.toList());
+                List<PackageNode> result = Arrays.stream(references)
+                        .filter(entry -> entry.getEntryKind() != IClasspathEntry.CPE_LIBRARY && entry.getEntryKind() != IClasspathEntry.CPE_VARIABLE)
+                        .map(entry -> getNodeFromClasspathEntry(entry, javaProject, NodeKind.CONTAINER)).filter(containerNode -> containerNode != null)
+                        .collect(Collectors.toList());
+                List<IClasspathEntry> referenceLibraries = Arrays.stream(references)
```
Comment (andxu) [inline jdtls.ext/com.microsoft.jdtls.ext.core/src/com/microsoft/jdtls/ext/core/PackageCommand.java]: bad perf deciding referenceLibraries.size > 0
use anyMatch

File: jdtls.ext/com.microsoft.jdtls.ext.core/src/com/microsoft/jdtls/ext/core/PackageCommand.java [hunk 10]
```diff
@@ -174,36 +189,85 @@
...
+     * @param classpathEntry
+     *            referenced cariable's classpath entry
+     * @return correspond package node
+     */
+    private static PackageNode getNodeFromClasspathVariable(IClasspathEntry classpathEntry) {
+        IClasspathEntry entry = JavaCore.getResolvedClasspathEntry(classpathEntry);
+        String name = classpathEntry.getPath().toPortableString();
+        String path = entry.getPath().toPortableString();
+        boolean isDirectory = new File(path).isDirectory();
+        return new PackageRootNode(name, path, NodeKind.PACKAGEROOT, isDirectory ? IPackageFragmentRoot.K_SOURCE : IPackageFragmentRoot.K_BINARY);
```
Comment (andxu) [inline jdtls.ext/com.microsoft.jdtls.ext.core/src/com/microsoft/jdtls/ext/core/PackageCommand.java]: K_BINARY

File: jdtls.ext/com.microsoft.jdtls.ext.core/src/com/microsoft/jdtls/ext/core/PackageCommand.java [hunk 11]
```diff
@@ -174,36 +188,87 @@
...
             }
         }
         return Collections.emptyList();
     }
 
+    /**
+     * Get container node of referenced libraries
+     */
+    private static ContainerNode getReferencedLibrariesContainerNode() {
+        return new ContainerNode(REFERENCED_LIBRARIES_CONTAINER_NAME, REFERENCED_LIBRARIES_PATH, NodeKind.CONTAINER, IClasspathEntry.CPE_CONTAINER);
```
Comment (andxu) [inline jdtls.ext/com.microsoft.jdtls.ext.core/src/com/microsoft/jdtls/ext/core/PackageCommand.java]: a static final field will do this well than a static method

**Pull request merged**